### DISCUSSION TITLE
Is there any way to support multitenancy with @RegisterRestClient via annotations/config.

### DISCUSSION THREAD

**Discussion by nitinty** [quarkusio/quarkus#41462] [Q&A]
We have a service which fetches the data from different clients. The request and response contracts are same for both the clients. But difference is the base url on which they exposes the apis to fetch the details.

Is there any way to call the specific client APIs based on the **tenantId** we are receiving in header from same RestClient.

TENANT_ONE  - Client 1 - url  - http://client1.com/data/
TENANT_TWO - Client 2 - url  - http://client2.com/data/

Please assist here.

[a comment by an automated account is not shown]

**Comment by nitinty** [comment]
@Ladicek  @geoand

**Reply by geoand** [reply]
In this case, I would use programmatic creation of the client with `QuarkusRestClientBuilder` instead of using `@RegisterRestClient`

**Comment by nitinty** [comment]
@geoand 
I hope you are talking about below approach, But the problem with this approach is 
1) We can't leverage the @RestClient to inject  client 
2) Every time we have to call the provider to provide the specific client.
 

```
@RequestScoped
public class MultiTenantRestClientProvider {
    @Inject
    MultiTenantService multiTenantService;
    public <T> T getClient(Class<T> clientClass, String tenantId) {
        TenantProperties properties = multiTenantService.getTenantProperties(tenantId);
        return QuarkusRestClientBuilder.newBuilder()
                .baseUri(URI.create(properties.baseUrl()))
                .build(clientClass);
    }
}



@ApplicationScoped
public class MultiTenantOrderService {
    /*@RestClient
    MultiTenantProductClient multiTenantProductClient;
     */
    public ProductDTO getOrderByTenant(String productId) {
        MultiTenantProductClient productClient = restClientProvider.getClient(MultiTenantProductClient.class);
        return productClient.getProductByTenant(productId);
        return multiTenantProductClient.getProductByTenant(productId);
    }
}
```

**Reply by nitinty** [reply]
Isn't that possible we enhance the functionality for @RegisterRestClient and start consuming the configuration in below format ? And at runtime let the client decide which remote service needs to be called.

<img width="601" alt="Screenshot 2024-08-10 at 11 58 38 AM" src="https://github.com/user-attachments/assets/046188df-a89b-4bd4-9863-9dc8a379b90f">

**Reply by geoand** [reply]
You can perhaps look into using `ClientRequestFilter` which allows you to set the URI of the request
